crates/services/txpool_v2/src/collision_manager/basic.rs

```diff
@@ -0,0 +1,225 @@
+                MessageDataPredicate,
+                MessageDataSigned,
+            },
+        },
+        BlobId,
+        ContractId,
+        Input,
+        Output,
+        UtxoId,
+    },
+    fuel_types::Nonce,
+    services::txpool::PoolTransaction,
+};
+
+use crate::{
+    error::{
+        CollisionReason,
+        Error,
+    },
+    storage::StorageData,
+};
+
+use super::{
+    CollisionManager,
+    Collisions,
+};
+
+pub struct BasicCollisionManager<StorageIndex> {
+    /// Message -> Transaction that currently use the Message
+    messages_spenders: HashMap<Nonce, StorageIndex>,
+    /// Coins -> Transaction that currently use the UTXO
+    coins_spenders: HashMap<UtxoId, StorageIndex>,
+    /// Contract -> Transaction that currently create the contract
+    contracts_creators: HashMap<ContractId, StorageIndex>,
+    /// Blob -> Transaction that currently create the blob
+    blobs_users: HashMap<BlobId, StorageIndex>,
+}
+
+impl<StorageIndex> BasicCollisionManager<StorageIndex> {
+    pub fn new() -> Self {
+        Self {
+            messages_spenders: HashMap::new(),
+            coins_spenders: HashMap::new(),
+            contracts_creators: HashMap::new(),
+            blobs_users: HashMap::new(),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn is_empty(&self) -> bool {
+        self.messages_spenders.is_empty()
+            && self.coins_spenders.is_empty()
+            && self.contracts_creators.is_empty()
+            && self.blobs_users.is_empty()
+    }
+}
+
+impl<StorageIndex> Default for BasicCollisionManager<StorageIndex> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<StorageIndex> CollisionManager for BasicCollisionManager<StorageIndex>
+where
+    StorageIndex: Copy + Debug + Hash + PartialEq + Eq,
+{
+    type StorageIndex = StorageIndex;
+
+    fn find_collisions(
+        &self,
+        transaction: &PoolTransaction,
+    ) -> Result<Collisions<Self::StorageIndex>, Error> {
+        let mut collisions = HashMap::new();
+        if let PoolTransaction::Blob(checked_tx, _) = &transaction {
+            let blob_id = checked_tx.transaction().blob_id();
+            if let Some(state) = self.blobs_users.get(blob_id) {
+                collisions.insert(*state, vec![CollisionReason::Blob(*blob_id)]);
+            }
+        }
+        for input in transaction.inputs() {
+            match input {
+                Input::CoinSigned(CoinSigned { utxo_id, .. })
+                | Input::CoinPredicate(CoinPredicate { utxo_id, .. }) => {
+                    // Check if the utxo is already spent by another transaction in the pool
+                    if let Some(storage_id) = self.coins_spenders.get(utxo_id) {
+                        let entry = collisions.entry(*storage_id).or_default();
+                        entry.push(CollisionReason::Utxo(*utxo_id));
+                    }
+                }
+                Input::MessageCoinSigned(MessageCoinSigned { nonce, .. })
```

Comment:
cc @xgreenx as i don't have the answer

Reply:
Because `MessageId` already is used in another context. The `MessageId` is a hash over all fields of the `Message` including the `Nonce`. But only `Nonce` makes this `MessageId` unique, so we use it for the database and inputs instead.

Reply:
Okay thanks for clarifying. Is `Nonce` guaranteed to be unique or is it user-controlled? I.e. could a user forge a message that collides with another message and have it evicted from the txpool?